### PR TITLE
[ONNX] Fix input sequence for pad op

### PR DESCRIPTION
**Pull request by neginraoof**
Fix input sequence type inference for pad op.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -6078,20 +6116,20 @@ def test_constant_pad(self):
...
             def forward(self, x, pad: List[int]):
                 return torch.nn.functional.pad(x, pad)
 
         x = torch.randn(2, 2, 4, 4)
-        y = pad = (torch.tensor(2, dtype=torch.int32), torch.tensor(4, dtype=torch.int32))
+        y = pad = [2, 4]
         self.run_test(Pad(), (x, y))
 
-        y = pad = (torch.tensor(2, dtype=torch.int64), torch.tensor(4, dtype=torch.int64))
+        y = pad = [torch.tensor(2, dtype=torch.int64), torch.tensor(4, dtype=torch.int64)]
```
Comment (jiafatom) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6129]: Can we still maintain the old case rather than replacing it?

Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:6129]: No, the point of this PR is that we now distinguish between tuples and lists, as we don't flatten list inputs anymore.
The model in this case expects a list as input (line 6122).

File: torch/onnx/symbolic_opset11.py [hunk 2]
```diff
@@ -435,6 +435,8 @@ def unbind(g, self, dim=0, _outputs=None):
 #          The order is dim_n_begin, dim_n_end, dim_n-1_begin, dim_n-1_end, ..., dim_m_begin, dim_m_end,
 #          where m is in range [0, n].
 def _prepare_onnx_paddings(g, dim, pad):
+    if not sym_help._is_packed_list(pad) and sym_help._is_list(pad) and sym_help._is_scalar_list(pad):
```
Comment (jiafatom) [inline torch/onnx/symbolic_opset11.py:438]: Does it well handle empty list case?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -37,11 +37,39 @@
...
+    elif isinstance(elem, bool):
+        return np.array(elem, dtype=bool)
+    elif isinstance(elem, int):
+        return np.array(elem, dtype=int)
+    elif isinstance(elem, float):
+        return np.array(elem, dtype=float)
+    elif isinstance(elem, dict):
+        dict_ = []
+        for k in elem:
+            dict_ += [to_numpy(k)] + [to_numpy(elem[k])]
```
Comment (shubhambhokare1) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:68]: Are we flattening both the key and value in the case of a dict?

File: torch/onnx/symbolic_opset11.py (hunk 2 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset11.py:438]: Yes, this works for empty tensors as well. 
Though ORT currently cannot handle empty input list.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 3 above)
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:68]: Yes, keys and values are flattened and concatenated.

**Comment by BowenBao** [comment]
is this PR depending on #60335?

**Comment by neginraoof** [comment]
> is this PR depending on #60335?

No, this is ready for review/merge

**Comment by neginraoof** [comment]
@BowenBao Please help review the PR. Thanks!

**Comment by jiafatom** [comment]
@BowenBao Could you please take a look? Thanks.

File: torch/csrc/jit/serialization/export.cpp [hunk 4]
```diff
@@ -333,21 +341,19 @@ EncoderBase::EncoderBase(
   model_proto_.set_producer_version(onnx_torch::PRODUCER_VERSION);
 }
 
-void EncoderBase::EncodeValueInfo(
-    onnx::GraphProto* graph_proto,
-    onnx::ValueInfoProto* v,
+void EncoderBase::EncodeValueInfoType(
```
Comment (BowenBao) [inline torch/csrc/jit/serialization/export.cpp:344]: Is this part only refactoring? or is there some logic changes?

File: torch/csrc/jit/python/script_init.cpp [hunk 5]
```diff
@@ -478,7 +478,8 @@ static void setInputTensorTypes(
     TORCH_INTERNAL_ASSERT(input_values.size() == param_count_list.size());
   }
   for (auto v : input_values) {
-    AT_ASSERT(s_iter != stack.end());
+    if (s_iter == stack.end())
```
Comment (BowenBao) [inline torch/csrc/jit/python/script_init.cpp:481]: why the mismatch now?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 3 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:68]: how does it work for dict? is there any current test case for it?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 6]
```diff
@@ -259,14 +295,15 @@ def run_test(self, model, input, rtol=1e-3, atol=1e-7, do_constant_folding=True,
                  batch_size=2, use_gpu=True, dynamic_axes=None, test_with_inputs=None,
                  input_names=None, output_names=None, fixed_batch_size=False, dict_check=True,
                  training=None, remained_onnx_input_idx=None):
-        def _run_test(m, remained_onnx_input_idx):
+        def _run_test(m, remained_onnx_input_idx, flatten=True):
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:298]: this `flatten` flag seem not used anywhere, is this needed?

Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:300]: It is used in line 175, maybe a rebase issue.

File: torch/csrc/jit/serialization/export.cpp (hunk 4 above)
Comment (neginraoof) [reply torch/csrc/jit/serialization/export.cpp:344]: To extend EncodeValueInfo to support more complex (nested) types, I refactored the part of the code that encodes type as a recursive function.

File: torch/csrc/jit/python/script_init.cpp (hunk 5 above)
Comment (neginraoof) [reply torch/csrc/jit/python/script_init.cpp]: Needed this for debug. Reverted. Thanks.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 3 above)
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:70]: So far we only test dict when tracing, and for those cases torch jit faltten pass is used to flatten input/outputs.
In those cases, the jit flatten pass creates a list of flattened keys and values of the dict.
But no, we don't have a test for actual Dict input/outputs (scripting) as those are not supported by ONNX. This part is added for completeness.

File: torch/csrc/jit/serialization/export.cpp [hunk 7]
```diff
@@ -333,21 +341,19 @@ EncoderBase::EncoderBase(
...
         std::unordered_map<int64_t, std::string>>& dynamic_axes) {
-  std::string name = n->debugName();
-  v->set_name(name);
-
-  auto tensorTypeToONNXType = [&dynamic_axes, &name, n, this](
+  auto tensorTypeToONNXType = [&dynamic_axes, n, this](
                                   const TensorTypePtr& t,
-                                  onnx::TypeProto_Tensor* tensor_type) {
+                                  onnx::TypeProto_Tensor* onnx_tensor_type) {
+    std::string name = n->debugName();
```
Comment (BowenBao) [inline torch/csrc/jit/serialization/export.cpp:354]: nit unused?

**Review by BowenBao — APPROVED** [review]

**Pull request merged**